### DISCUSSION TITLE
Tortoise-Orm and Sqlalchemy are blocking and synchronous.

### DISCUSSION THREAD

**Discussion by Randix6644** [fastapi/fastapi#9354] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@app.get("/serversv2", response_model=ServerList) // sqlalchemy
async def get_all_servers(session: AsyncSession = Depends(get_session)):
    result = await session.execute(select(Server1))
    servers = result.scalars().all()
    return {"servers": servers}


@app.get("/servers") // tortoise-orm
async def get_servers():
    return await Server.all()
```

### Description

Using fastapi with tortoise-orm==0.19.3. There're like 25,000 records of the table `Server`.

I'm fetching all of them with the endpoint `/servers/` like this.
```python
@server_router.get("/servers")
async def get_servers():
    ret = await ServerPydantic.from_queryset(Server.all())
    return ret
```
There's no pagination and offset, I do this for test purpose.

There is another endpoint:
```python
@server_router.get("/servers-test/{s}")
async def test(s:int):
    print(f"sleeping for {s} seconds")
    await asyncio.sleep(s)
    return {"test": "test"}
```
I make requests to the endpoint`/server`, then I quickly make two more requests with the endpoints `/servers-test/{s}`:
- `curl http://127.0.0.1/servers` , this will take some time, around 4-5s
- `curl http://127.0.0.1/servers-test/5`, right after the first request
- `curl http://127.0.0.1/servers-test/0` right after the previous request

I can tell immediately by looking at the print statement in the functions the last requests are not processed at all, it's blocking by the call to `/servers` until the first request finishes. Only then the other two start to run.

If i do
```python
@server_router.get("/servers")
async def get_servers():
    asyncio.sleep(30)
    print("sleeping done")
    ret = await ServerPydantic.from_queryset(Server.all())
    return ret
```
Basically I just added the `asyncio.sleep(30)`. Then I repeat the steps mentioned above.

- `curl http://127.0.0.1/servers` , this will take even longer time since `asyncio.sleep(30)` is added

- `curl http://127.0.0.1/servers-test/5`, right after the first request, it gets served right away
- `curl http://127.0.0.1/servers-test/0` right after the previous request, it gets served immediately

If I make a request after 30 seconds, then requests get blocked again. So obviously, the `Server.all()` blocks entire thread. It's not expected. Am I doing anything wrong? I've searched on Google and Tortoise-orm Github issues, and yet there's no related info found.

Then I tried sqlalchemy like so:

```python
#  some basic setups are not shown here.

@app.get("/serversv2", response_model=ServerList) # sqlalchemy
async def get_all_servers(session: AsyncSession = Depends(get_session)):
    result = await session.execute(select(Server1))
    servers = result.scalars().all()
    return {"servers": servers}
```

And the same thing happens. requests to `servers-test` gets blocked.

### Operating System

macOS

### FastAPI Version

0.95.0

### Python Version

Python 3.8.2

**Comment by yinziyan1206** [comment]
I use `SQLAlchemy` and i don't have such problem. I guess the blocking-point is when it serializes data to row. You could try data stream in `SQLAlchemy` and test it if it blocks.

**Reply by Randix6644** [reply]
Honestly, I'm new to Sqlalchemy.  But I'll look into that. Serializing or deserializing data should happen in memory , shouldn't it?  Any idea how I can debug this?

**Reply by yinziyan1206** [reply]
you can use data stream and async iterator. Then you can serialize one by one to see if it really blocks

**Comment by vsdudakov** [comment]
Small correction on the premise: both Tortoise and SQLAlchemy 2.0 have genuinely async modes now (asyncpg / tokio underneath), so they're not blocking per se - the async win just only shows up under concurrent I/O, not on a single sequential query. 

If what you're really after is a Tortoise-style async API with a faster non-blocking engine, I maintain Yara ORM: the hot path (pooling, binding, row decode) is in Rust via PyO3 (tokio-postgres / rusqlite), Postgres + SQLite, prebuilt wheels so there's no Rust toolchain to install. Most Tortoise code moves over unchanged.
(Disclosure: I'm the author - [github.com/vsdudakov/yara-orm](https://github.com/vsdudakov/yara-orm))
